oauth/authcode.go

```diff
@@ -0,0 +1,183 @@
+// query param when it gets a request.
+type authHandler struct {
+	c chan string
+}
+
+func (h authHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.c <- r.URL.Query().Get("code")
+	w.Header().Set("Content-Type", "text/html")
+	w.Write([]byte("<html><body><p>Login successful. Please return to the terminal. You may now close this window.</p></body></html>"))
+}
+
+// AuthorizationCodeTokenSource with PKCE as described in:
+// https://www.oauth.com/oauth2-servers/pkce/
+// This works by running a local HTTP server on port 8484 and then having the
+// user log in through a web browser, which redirects to the local server with
+// an authorization code. That code is then used to make another HTTP request
+// to fetch an auth token (and refresh token). That token is then in turn
+// used to make requests against the API.
+type AuthorizationCodeTokenSource struct {
+	ClientID       string
+	AuthorizeURL   string
+	TokenURL       string
+	EndpointParams *url.Values
+	Scopes         []string
+}
+
+// Token generates a new token using an authorization code.
+func (ac *AuthorizationCodeTokenSource) Token() (*oauth2.Token, error) {
+	// Generate a random code verifier string
+	verifierBytes := make([]byte, 32)
+	if _, err := rand.Read(verifierBytes); err != nil {
+		return nil, err
+	}
+
+	verifier := base64.RawURLEncoding.EncodeToString(verifierBytes)
+
+	// Generate a code challenge. Only the challenge is sent when requesting a
+	// code which allows us to keep it secret for now.
+	shaBytes := sha256.Sum256([]byte(verifier))
+	challenge := base64.RawURLEncoding.EncodeToString(shaBytes[:])
+
+	// Generate a URL with the challenge to have the user log in.
+	url := fmt.Sprintf("%s?response_type=code&code_challenge=%s&code_challenge_method=S256&client_id=%s&redirect_uri=http://localhost:8484/&scope=%s", ac.AuthorizeURL, challenge, ac.ClientID, strings.Join(ac.Scopes, `%20`))
+
+	if len(*ac.EndpointParams) > 0 {
+		url += "&" + ac.EndpointParams.Encode()
+	}
+
+	// Run server before opening the user's browser so we are ready for any redirect.
+	codeChan := make(chan string)
+	handler := authHandler{
+		c: codeChan,
+	}
+
+	s := &http.Server{
+		Addr:           ":8484",
+		Handler:        handler,
+		ReadTimeout:    5 * time.Second,
+		WriteTimeout:   5 * time.Second,
+		MaxHeaderBytes: 1024,
+	}
+
+	go func() {
+		// Run in a goroutine until the server is closed or we get an error.
+		if err := s.ListenAndServe(); err != http.ErrServerClosed {
+			log.Fatal().Err(err).Msg("Server exited unexpectedly")
+		}
+	}()
+
+	// Open auth URL in browser, print for manual use in case open fails.
+	fmt.Println("Open your browser to log in using the URL:")
+	fmt.Println(url)
+	open(url)
+
+	// Get code from handler, exchange it for a token, and then return it. This
+	// channel read blocks until one code becomes available.
+	// There is currently no timeout.
+	code := <-codeChan
+	s.Shutdown(context.Background())
```

Comment:
We might also want to print something out here that lets the user know that the login was successful. Especially since the last message on the console will be instructions to open the browser.

Reply:
I opted not to do this because in practice with my sample CLI it just continues and shows the output of the command. You can always pass `--verbose` and see additional info about the request being made with the newly fetched token so I'm not sure there's a ton of value adding an additional message.